File: Saikat/SVM.py
import numpy as np

class SVM:
    def __init__(self, C=1.0, tol=0.001, max_passes=5):
        self.C = C
        self.tol = tol
        self.max_passes = max_passes

    def fit(self, X, y):
        m, n = X.shape
        self.alphas = np.zeros(m)
        self.b = 0
        self.X = X
        self.y = y
        passes = 0

        while passes < self.max_passes:
            num_changed_alphas = 0
            for i in range(m):
                Ei = self._decision_function(X[i]) - y[i]
                if (y[i] * Ei < -self.tol and self.alphas[i] < self.C) or (y[i] * Ei > self.tol and self.alphas[i] > 0):
                    j = self._select_j(i, m)
                    Ej = self._decision_function(X[j]) - y[j]

                    old_alpha_i = self.alphas[i]
                    old_alpha_j = self.alphas[j]

                    if y[i] != y[j]:
                        L = max(0, self.alphas[j] - self.alphas[i])
                        H = min(self.C, self.C + self.alphas[j] - self.alphas[i])
                    else:
                        L = max(0, self.alphas[i] + self.alphas[j] - self.C)
                        H = min(self.C, self.alphas[i] + self.alphas[j])

                    if L == H:
                        continue

                    eta = 2.0 * self._kernel(X[i], X[j]) - self._kernel(X[i], X[i]) - self._kernel(X[j], X[j])
                    if eta >= 0:
                        continue

                    self.alphas[j] -= y[j] * (Ei - Ej) / eta
                    self.alphas[j] = self._clip_alpha(self.alphas[j], H, L)

                    if abs(self.alphas[j] - old_alpha_j) < 1e-5:
                        continue

                    self.alphas[i] += y[i] * y[j] * (old_alpha_j - self.alphas[j])

                    b1 = self.b - Ei - y[i] * (self.alphas[i] - old_alpha_i) * self._kernel(X[i], X[i]) - y[j] * (self.alphas[j] - old_alpha_j) * self._kernel(X[i], X[j])
                    b2 = self.b - Ej - y[i] * (self.alphas[i] - old_alpha_i) * self._kernel(X[i], X[j]) - y[j] * (self.alphas[j] - old_alpha_j) * self._kernel(X[j], X[j])

                    if 0 < self.alphas[i] < self.C:
                        self.b = b1
                    elif 0 < self.alphas[j] < self.C:
                        self.b = b2
                    else:
                        self.b = (b1 + b2) / 2

                    num_changed_alphas += 1

            if num_changed_alphas == 0:
                passes += 1
            else:
                passes = 0

    def predict(self, X):
        return np.sign(self._decision_function(X))

    def _decision_function(self, X):
        result = 0
        for i in range(len(self.X)):
            result += self.alphas[i] * self.y[i] * self._kernel(self.X[i], X)
        return result + self.b

    def _kernel(self, x1, x2):
        return np.dot(x1, x2)

    def _select_j(self, i, m):
        j = i
        while j == i:
            j = np.random.randint(0, m)
        return j

    def _clip_alpha(self, alpha, H, L):
        if alpha > H:
            return H
        elif alpha < L:
            return L
        else:
            return alpha

if __name__ == "__main__":
    X = np.array([
        [1, 2],
        [2, 3],
        [3, 3],
        [2, 1],
        [3, 2]
    ])
    y = np.array([1, 1, 1, -1, -1])

    svm = SVM(C=1.0, tol=0.001, max_passes=1000)
    svm.fit(X, y)

    test_point = np.array([2.5, 2.5])
    prediction = svm.predict(test_point)
    print("Prediction for [2.5, 2.5]:", prediction)

File: Saikat/SVM.java
import java.util.Random;

public class SVM {
    private double C;
    private double tol;
    private int maxPasses;
    private double[] alphas;
    private double b;
    private double[][] X;
    private double[] y;
    private Random random;

    public SVM(double C, double tol, int maxPasses) {
        this.C = C;
        this.tol = tol;
        this.maxPasses = maxPasses;
        this.random = new Random();
    }

    public void fit(double[][] X, double[] y) {
        int m = X.length;
        int n = X[0].length;
        this.alphas = new double[m];
        this.b = 0;
        this.X = X;
        this.y = y;
        int passes = 0;

        while (passes < this.maxPasses) {
            int numChangedAlphas = 0;
            for (int i = 0; i < m; i++) {
                double Ei = decisionFunction(X[i]) - y[i];
                if ((y[i] * Ei < -this.tol && this.alphas[i] < this.C) || (y[i] * Ei > this.tol && this.alphas[i] > 0)) {
                    int j = selectJ(i, m);
                    double Ej = decisionFunction(X[j]) - y[j];

                    double oldAlphaI = this.alphas[i];
                    double oldAlphaJ = this.alphas[j];

                    double L, H;
                    if (y[i] != y[j]) {
                        L = Math.max(0, this.alphas[j] - this.alphas[i]);
                        H = Math.min(this.C, this.C + this.alphas[j] - this.alphas[i]);
                    } else {
                        L = Math.max(0, this.alphas[i] + this.alphas[j] - this.C);
                        H = Math.min(this.C, this.alphas[i] + this.alphas[j]);
                    }

                    if (L == H) continue;

                    double eta = 2.0 * kernel(X[i], X[j]) - kernel(X[i], X[i]) - kernel(X[j], X[j]);
                    if (eta >= 0) continue;

                    this.alphas[j] -= y[j] * (Ei - Ej) / eta;
                    this.alphas[j] = clipAlpha(this.alphas[j], H, L);

                    if (Math.abs(this.alphas[j] - oldAlphaJ) < 1e-5) continue;

                    this.alphas[i] += y[i] * y[j] * (oldAlphaJ - this.alphas[j]);

                    double b1 = this.b - Ei - y[i] * (this.alphas[i] - oldAlphaI) * kernel(X[i], X[i]) - y[j] * (this.alphas[j] - oldAlphaJ) * kernel(X[i], X[j]);
                    double b2 = this.b - Ej - y[i] * (this.alphas[i] - oldAlphaI) * kernel(X[i], X[j]) - y[j] * (this.alphas[j] - oldAlphaJ) * kernel(X[j], X[j]);

                    if (0 < this.alphas[i] && this.alphas[i] < this.C) {
                        this.b = b1;
                    } else if (0 < this.alphas[j] && this.alphas[j] < this.C) {
                        this.b = b2;
                    } else {
                        this.b = (b1 + b2) / 2.0;
                    }

                    numChangedAlphas++;
                }
            }

            if (numChangedAlphas == 0) {
                passes++;
            } else {
                passes = 0;
            }
        }
    }

    public double predict(double[] X) {
        return Math.signum(decisionFunction(X));
    }

    private double decisionFunction(double[] X) {
        double result = 0;
        for (int i = 0; i < this.X.length; i++) {
            result += this.alphas[i] * this.y[i] * kernel(this.X[i], X);
        }
        return result + this.b;
    }

    private double kernel(double[] x1, double[] x2) {
        double sum = 0;
        for (int i = 0; i < x1.length; i++) {
            sum += x1[i] * x2[i];
        }
        return sum;
    }

    private int selectJ(int i, int m) {
        int j = i;
        while (j == i) {
            j = this.random.nextInt(m);
        }
        return j;
    }

    private double clipAlpha(double alpha, double H, double L) {
        if (alpha > H) {
            return H;
        } else if (alpha < L) {
            return L;
        } else {
            return alpha;
        }
    }

    public static void main(String[] args) {
        double[][] X = {
            {1, 2},
            {2, 3},
            {3, 3},
            {2, 1},
            {3, 2}
        };
        double[] y = {1, 1, 1, -1, -1};

        SVM svm = new SVM(1.0, 0.001, 1000);
        svm.fit(X, y);

        double[] testPoint = {2.5, 2.5};
        double prediction = svm.predict(testPoint);
        System.out.println("Prediction for [2.5, 2.5]: " + prediction);
    }
}


